### ISSUE TITLE
ENH Add "adversarial debiasing" to mitigations

### ISSUE THREAD

**Issue by romanlutz** [fairlearn/fairlearn#785]
The paper [Mitigating Unwanted Biases with Adversarial Learning](https://arxiv.org/abs/1801.07593) by Zhang, Lemoine, and Mitchell presents an "adversarial debiasing" technique. The goal of this task is to implement it.

AIF360 has an [implementation](https://aif360.readthedocs.io/en/v0.2.3/modules/inprocessing.html#aif360.algorithms.inprocessing.AdversarialDebiasing) of this technique, but it uses `tensorflow`. Fairlearn shouldn't depend on `tensorflow` (or `pytorch`, for that matter), but instead have a soft dependency on the package of choice. @adrinjalali has previously noted that Fairlearn should prefer `pytorch` over `tensorflow`, which we should take into account for this.

[Note: The name "adversarial debiasing" is in conflict with [our policy](https://fairlearn.org/main/contributor_guide/how_to_talk_about_fairness.html) around the words "bias"/"debiasing". What else could we call it? Adversarial fairness? @fairlearn/fairlearn-maintainers ]

Completing this item requires:
- code for the technique in `fairlearn.adversarial` (@fairlearn/fairlearn-maintainers  is that a suitable module name? Other suggestions?)
- unit tests in `test.unit.adversarial`
- descriptive API reference (directly in the docstring)
- ideally a short user guide in docs.user_guide.mitigation.rst

A fully fledged example notebook is not required.

To claim this task please respond below. Of course, you can also ask questions!

**Comment by adrinjalali** [comment]
I'd be a bit concerned with our policy of not using the word "bias" or "debias" completely. These words are used in the literature and folks use them to find resources and implementations. I completely agree that we can't completely debias a dataset, or a model, but we do remote some sort of a bias when we use these methods. We can still use these words, but be very clear about what kind of bias it is that the method is removing, instead of saying this method debiases the model/data.

**Comment by romanlutz** [comment]
Exactly, but "Adversarial Debiasing" isn't a clear name since it misses exactly that point. In fact, it might give the impression that it's a fix for all biases.

Thanks @riedgar-ms for the title fix!

**Comment by hildeweerts** [comment]
Within the paper itself they even acknowledge that debiasing might be problematic:*"“debias” may not be quite the right word, as all bias is not necessarily removed"*. 

I agree with @adrinjalali that in some cases the usage of "bias" or "de-bias" might be appropriate, but I wouldn't say that's the case for this specific paper. In particular, there is no exact specification of the type of "bias" that is "removed" and where it is supposed to originate from. Instead, like most fairness mitigation approaches, the method optimizes for a particular fairness metrics, foregoing the biases from which potential unfairness may have originated. 

My first thought was also to go for something like "AdversarialFairness" - although that also still sounds a bit absolute and that it's related to fairness is probably clear from the fact that it's part of Fairlearn. "AdversarialMitigation" could be an option.

But I also share @adrinjalali 's concern in terms of the discoverability of the approach...

@fairlearn/fairlearn-maintainers thoughts?

**Comment by stephenrobic** [comment]
I will begin working on this issue, as I have some spare time and this seems like a doable step up from #812.

**Comment by hildeweerts** [comment]
Hi @stephenrobic, did you have any luck trying to implement this? Please let us know if you have any questions!

If you don't plan to work on this issue anymore that is also perfectly fine. Do let us know so we can let other people know the issue is up for grabs again!

**Comment by hildeweerts** [comment]
Hi @stephenrobic! Please let me know if you'd still like to work on this issue!

**Comment by SeanMcCarren** [comment]
I would like to work on this issue :)

**Comment by stephenrobic** [comment]
Please take and work on this issue! I thought I was prepared to work on
this, but I haven't been able to.

**Comment by SeanMcCarren** [comment]
Hi there, I have some questions regarding this implementation. The paper leaves room for quite some freedom, and we need to make some tough decision choices regarding this. Most notably, I can think of situations in which one may want to change:
- Predictor/adversarial models (make predictor very complex, possibly)
- Output layer. Continuous vs discrete (and for discrete with demographic parity, single binary->sigmoid vs multiclass->cross entropy is another choice). Loss functions depend on this.
- Which metric. The paper speaks about DP, Equality of odds, Equality of opportunity, but they require slightly different scenarios (Equality of odds requires adversary has more inputs and binary values, whereas Equality of opportunity requires predictions are discrete)

One possible implementation is to provide the required code that one would put in a training loop, but not more. Then, a user of this module has to implement their own training loop. Pros: lots of freedom, can do validation at every step, batching how the user would like. Cons: users will have to know precisely what parameters are possible, and must understand the original paper or this function really well.

Another possibility, what I see in AIF360 (among others) also, is to provide the training loop and most choices given above as standard. Pros: Simpler use, maybe we can even use the standard `.fit(X,y)` and `.predict(X)` as the user needn't care how it is fit (what parameters/training loop.) which fits in well with sklearn-style. We can define the different metrics, the user just has to choose. Cons: no custom training loops, so besides possibly some keyword parameters the user cant add things like validation every X epochs or batching, neither can the user define custom adversary models. We don't know how much to train (but the user can supply epochs or something at least)

Side note: Where can I write in the user guide for this? I was thinking under mitigations, but those are about algorithms that implement `.fit(X,y)` and `.predict(X)`, and I don't know whether this algorithm will have this.

**Comment by taharallouche** [comment]
I suspect close this one as well, given https://github.com/fairlearn/fairlearn/pull/1079 was merged a while ago

**Comment by hildeweerts** [comment]
Agreed, @taharallouche!
